### PR TITLE
Add recomputation check for number representation.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libevmasm/ConstantOptimiser.h [hunk 1]
```diff
@@ -130,6 +133,8 @@ class ComputeMethod: public ConstantOptimisationMethod
 		ConstantOptimisationMethod(_params, _value)
 	{
 		m_routine = findRepresentation(m_value);
+		if (!checkRepresentation(m_value))
+			BOOST_THROW_EXCEPTION(AssemblyException());
```
Comment (axic) [inline libevmasm/ConstantOptimiser.h]: Apparently we have `OptimizerException` too 😉

Comment (axic) [reply libevmasm/ConstantOptimiser.h]: Also can we use `assertThrow(checkRepresentation(m_value), OptimizerException, "Invalid constant expression created.");` ?

File: libevmasm/ConstantOptimiser.h [hunk 2]
```diff
@@ -141,6 +146,8 @@ class ComputeMethod: public ConstantOptimisationMethod
 protected:
 	/// Tries to recursively find a way to compute @a _value.
 	AssemblyItems findRepresentation(u256 const& _value);
+	/// Recomputes the value from the representation and checks for correctness.
+	bool checkRepresentation(u256 const&);
```
Comment (axic) [inline libevmasm/ConstantOptimiser.h]: For clarity it may make sense to actually have `AssemblyItems` as a parameter here and not use the member directly.

Comment (axic) [reply libevmasm/ConstantOptimiser.h]: Also why not have this parameter named `_value`?

Comment (chriseth) [reply libevmasm/ConstantOptimiser.h]: All the other functions here are stateful, too, so I think it is fine to use the member, especially for a protected function. Will add the parameter name, though.

**Review by axic — APPROVED** [review]
Looks good.

**Comment by pirapira** [comment]
I don't review this because it is already merged.

**Pull request merged**